aten/src/ATen/core/DistributionsHelper.h

```diff
@@ -0,0 +1,288 @@
+#pragma once
+
+// define constants like M_PI and C keywords for MSVC
+#ifdef _MSC_VER
+#define _USE_MATH_DEFINES
+#include <math.h>
+#endif
+
+#include <ATen/CPUGenerator.h>
+#include <ATen/core/Array.h>
+#include <type_traits>
+#include <limits>
+#include <cmath>
+
+/** 
+ * Distributions kernel adapted from THRandom.cpp
+ * The kernels try to follow std::random distributions signature
+ * For instance: in ATen
+ *      auto gen = at::detail::createCPUGenerator();
+ *      at::uniform_real_distribution<double> uniform(0, 1);
+ *      auto sample = uniform(gen.get());
+ *      
+ *      vs std::random
+ * 
+ *      std::mt19937 gen;
+ *      std::uniform_real_distribution uniform(0, 1);
+ *      auto sample = uniform(gen);
+ */
+
+
+namespace at {
+
+// Using VectorType in Box-muller derived distributions to avoid
+// code duplication
+template <typename T>
+struct VectorType {  };
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+template <> struct VectorType<half> { using type = at::detail::Array<float, 2>; };
+#endif
+template <> struct VectorType<Half> { using type = at::detail::Array<float, 2>; };
+template <> struct VectorType<float> { using type = at::detail::Array<float, 2>; };
+template <> struct VectorType<double> { using type = at::detail::Array<double, 2>; };
+
+template <typename T>
+using vect_type = typename VectorType<T>::type;
+
+// Using DistAccumType in accumulate types for distributions.
+// Note: Ideally we'd be using ATen/AccumulateType.h but looks
+// like the there is some inconsistency in how accumulate types
+// are mapped currently, e.g. for the cpu side, float is mapped
+// to double.
+template <typename T>
+struct DistAccumType {  };
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+template <> struct DistAccumType<half> { using type = float; };
+#endif
+template <> struct DistAccumType<Half> { using type = float; };
+template <> struct DistAccumType<float> { using type = float; };
+template <> struct DistAccumType<double> { using type = double; };
+
+template <typename T>
+using dist_acctype = typename DistAccumType<T>::type;
+
+// Constants for uniform distribution
+// doubles have 52 bits of mantissa (fractional part)
+constexpr uint64_t DOUBLE_MASK = (1ULL << 53) - 1;
+constexpr double DOUBLE_DIVISOR = 1.0 / (1ULL << 53);
+
+// floats have 23 bits of mantissa (fractional part)
+constexpr uint32_t FLOAT_MASK = (1 << 24) - 1;
+constexpr float FLOAT_DIVISOR = 1.0f / (1 << 24);
+
+/**
+ * Samples a uniform distribution in the range [0,1) of type T
+ */
+template <typename T>
+struct uniform_real_distribution {
+
+  C10_HOST_DEVICE inline uniform_real_distribution(T a_in, T b_in) {
```

Comment:
The error checking here being host+device supported feels goofy. Aren't we only ever going to do this error checking on CPU? Much better to avoid `assert()` in this case. I haven't audited the call sites to see if this is true.

Reply:
A brief audit of call sites suggests this is only called from CPU sites (of course, this may be because this refactor only handles CPU generators...)